In [73]:
from lightfm import LightFM, cross_validation
from lightfm.data import Dataset
from scipy.sparse import coo_matrix
from sklearn.externals import joblib
from lightfm.evaluation import auc_score
import json
import requests
import pandas as pd
from lightfm import LightFM, cross_validation
from lightfm.data import Dataset
from scipy.sparse import coo_matrix
from sklearn.externals import joblib
from lightfm.evaluation import auc_score

from IPython.display import Markdown as md

fr=2 #GHz

md("$f_r = %i$ GHz"%(fr))

data = pd.read_csv('write.csv')

In [8]:
from IPython.display import HTML

HTML('''<script>
code_show=true; 
function code_toggle() {
 if (code_show){
 $('div.input').hide();
 } else {
 $('div.input').show();
 }
 code_show = !code_show
} 
$( document ).ready(code_toggle);
</script>
The raw code for this IPython notebook is by default hidden for easier reading.
To toggle on/off the raw code, click <a href="javascript:code_toggle()">here</a>.''')

![title](card.svg)

In [15]:
viewsSum = data[data.actionCategory=='WebNei clicked'].groupby(['idVisit', 'actionName']).size().apply(lambda x: 1).sum()
users = data.userName.nunique()
presentations = data[data.actionCategory=='WebNei clicked'].actionName.nunique()

print('There are %s users.' %users)
print('There are %s presentations.' %presentations)
print('There are %s presentations views.' %viewsSum)

There are 7408 users.
There are 1489 presentations.
There are 102065 presentations views.


#### WebNei is a web application that stores technical presentations of Nokia Company.
#### It allows Nokia employees to browse, search and add presentations.

#### We had received data about users interactions with that application.

#### There are 7408 users, 1489 presentations and 102065 presentations views.
#### We have been collecting data for 550 days.


In [17]:
data.tail(8)

idVisit           serverTime  timeSpent  \
866693   662412  2018-05-16 06:34:20        NaN   
866694   662412  2018-05-16 06:34:20        3.0   
866695   662412  2018-05-16 06:34:23        NaN   
866696   662412  2018-05-16 06:34:23        NaN   
866697   662438  2018-05-16 06:34:32        9.0   
866698   662438  2018-05-16 06:34:41       14.0   
866699   662447  2018-05-16 06:35:03       26.0   
866700   662447  2018-05-16 06:35:29        NaN   

                        actionCategory  \
866693               Webnei downloaded   
866694             Category downloaded   
866695               Webnei downloaded   
866696             Category downloaded   
866697  CategoryChip filtering Clicked   
866698  CategoryChip filtering Clicked   
866699          Read document backward   
866700          Read document backward   

                                              actionName  \
866693  SR001503 Common Element Manager for LTE and SRAN   
866694                             SINGLE RAN downloaded   
866695  SR001503 Common Element Manager for LTE and SRAN   
866696                             SINGLE RAN downloaded   
866697                                             WCDMA   
866698                                        SINGLE RAN   
866699                           Intra eNB UL CoMP (2RX)   
866700                           Intra eNB UL CoMP (2RX)   

                              userName              department  \
866693      Awakening: Moonfell Wood    CO Customer Operations   
866694      Awakening: Moonfell Wood    CO Customer Operations   
866695      Awakening: Moonfell Wood    CO Customer Operations   
866696      Awakening: Moonfell Wood    CO Customer Operations   
866697                             NaN      GS Global Services   
866698                             NaN      GS Global Services   
866699  Bill Walsh College Football         MN Mobile Networks   
866700  Bill Walsh College Football         MN Mobile Networks   

                                    team                         title  \
866693  CO MEA MS RSO NPO CEWA &amp; SAV                  NPO Engineer   
866694  CO MEA MS RSO NPO CEWA &amp; SAV                  NPO Engineer   
866695  CO MEA MS RSO NPO CEWA &amp; SAV                  NPO Engineer   
866696  CO MEA MS RSO NPO CEWA &amp; SAV                  NPO Engineer   
866697          GS GSD C EUR RS Russia 2                      Engineer   
866698          GS GSD C EUR RS Russia 2                      Engineer   
866699     MN CDS Pz Radio Sys Perf PET1  Engineer, System Performance   
866700     MN CDS Pz Radio Sys Perf PET1  Engineer, System Performance   

                       organization                            name  \
866693  CO Middle East &amp; Africa      Awakening: Moonfell Wood     
866694  CO Middle East &amp; Africa      Awakening: Moonfell Wood     
866695  CO Middle East &amp; Africa      Awakening: Moonfell Wood     
866696  CO Middle East &amp; Africa      Awakening: Moonfell Wood     
866697   GS Global Service Delivery                             NaN   
866698   GS Global Service Delivery                             NaN   
866699                       MN CDS  Bill Walsh College Football      
866700                       MN CDS  Bill Walsh College Football      

             country  
866693         Kenya  
866694         Kenya  
866695         Kenya  
866696         Kenya  
866697  Russian Fed.  
866698  Russian Fed.  
866699       Finland  
866700       Finland

### We decided to provide WebNei users recommendations based on their view history, 
### their profession and membership to team and department.

### Recommender system is system that tries to find the best suitable item for each user.




### Kinds of recommender systems:
* #### CF - collaborative filtering- consist of predicting what user will like based on their similarity to other users
<img src="image1.jpg" alt="drawing" width="600px"/>


### Kinds of recommender systems:
* #### CB - content based- select items based on the features like tags or description of previous items seen by user
<img src="image2.jpg" alt="drawing" width="600px"/>


![sparse](sparse.svg)

![coldstart](coldstart.svg)

![whylightfm](whylightfm.svg)

![auc](auc.svg)

##  API usage
* #### Upload data, tags and compute model
#### `/admin/computeModelFromCsvWithTags`
##### Parameters
* ##### `data_file` - csv file with implicit data.
* ##### `tags_file` - json file with items features.
##### Output
* ##### If successful: _Sucesfully computed recommender model_.
* ##### Otherwise: _Message with error_.



##  API usage
* #### Get recommendations for user
#### `/recommendation/<userName>`
##### Parameters
* ##### None
##### Output
* ##### List of top 20 recommended presentations.
* ##### List of rank for each presentation.
* ##### Username.




## Examples

In [71]:
files = {'data_file': open('write.csv','rb'), 'tags_file': open('tagi.json','rb')}
r = requests.post("http://127.0.0.1:5000/admin/computeModelFromCsvWithTags", files=files)
r.text  

'Sucesfully computed recommender model'

## Examples

In [74]:
response = requests.get("http://127.0.0.1:5000/recommendation/Aigina's Prophecy  ")

js = response.json()
print(json.dumps(js, indent=2))

{
  "items": [
    "5G18A Transport Fundamentals",
    "5G17 Transport Complex",
    "BSS101417",
    "LTE1074",
    "LTE1067",
    "5GC000202",
    "5GC000548",
    "LTE1534",
    "5GC000078",
    "LTE1089",
    "LTE1529",
    "5GC000008",
    "BSS101428",
    "5GC000085",
    "LTE1486",
    "IMS Dimensioning Basics",
    "5G17 NB Architecture and HW Components Complex",
    "LTE1543",
    "AirScale RNC 17 Dimensioning Guideline",
    "5GC000547"
  ],
  "user": "Aigina's Prophecy  ",
  "weights": [
    2.916257858276367,
    2.5929694175720215,
    2.2521042823791504,
    1.999570608139038,
    1.9680728912353516,
    1.8758412599563599,
    1.713552713394165,
    1.6287225484848022,
    1.5843499898910522,
    1.515429973602295,
    1.4628912210464478,
    1.4245799779891968,
    1.4097068309783936,
    1.392263412475586,
    1.3919960260391235,
    1.3785371780395508,
    1.367828369140625,
    1.3635002374649048,
    1.282402753829956,
    1.2402098178863525
  ]
}


## Accuracy of recommendation

In [77]:
#Here We will show connections between recomendation for 'Aigina's Prophecy ' and his activity (a lot of lte and 5g stuff)
data[data.userName == "Aigina's Prophecy  "]

idVisit           serverTime  timeSpent              actionCategory  \
1706     202754  2016-12-20 08:14:50        4.0              WebNei clicked   
1707     202754  2016-12-20 08:14:54        NaN           Webnei downloaded   
1883     203299  2016-12-20 14:00:41       66.0              WebNei clicked   
1884     203299  2016-12-20 14:01:47        NaN           Webnei downloaded   
3268     205883  2016-12-23 08:41:32       50.0       Tag filtering Clicked   
3269     205883  2016-12-23 08:42:22        NaN              WebNei clicked   
3891     208356  2016-12-30 08:46:09      248.0              WebNei clicked   
3894     208356  2016-12-30 08:53:57       65.0              WebNei clicked   
11756    224811  2017-01-24 11:25:49        7.0  Category filtering Clicked   
11757    224811  2017-01-24 11:25:56       39.0  Category filtering Clicked   
11758    224811  2017-01-24 11:26:54        NaN  Category filtering Clicked   
14395    230723  2017-01-31 15:43:39       19.0       Tag filtering Clicked   
14396    230723  2017-01-31 15:43:58        NaN              WebNei clicked   
17126    236577  2017-02-07 13:32:09        NaN              WebNei clicked   
20890    242286  2017-02-14 09:53:50      514.0              WebNei clicked   
20906    242286  2017-02-14 10:02:29      196.0              WebNei clicked   
20922    242286  2017-02-14 10:18:36       43.0       Tag filtering Clicked   
20923    242286  2017-02-14 10:19:19      100.0              WebNei clicked   
20924    242286  2017-02-14 10:20:59      110.0       Tag filtering Clicked   
20925    242286  2017-02-14 10:22:49        NaN              WebNei clicked   
21611    243598  2017-02-15 11:52:07        NaN              WebNei clicked   
22223    244731  2017-02-16 11:55:59        NaN              WebNei clicked   
22244    244806  2017-02-16 12:51:29        NaN              WebNei clicked   
22353    245053  2017-02-16 15:53:18      468.0              WebNei clicked   
22357    245053  2017-02-16 16:01:06        NaN              WebNei clicked   
22943    246141  2017-02-17 16:29:23        NaN  Category filtering Clicked   
27688    254790  2017-03-01 10:07:36       25.0              WebNei clicked   
27689    254790  2017-03-01 10:08:40        NaN              WebNei clicked   
28182    255794  2017-03-02 08:38:18       21.0              WebNei clicked   
28183    255794  2017-03-02 08:38:39        NaN           Webnei downloaded   
...         ...                  ...        ...                         ...   
738287   641105  2018-04-30 07:58:34       25.0              WebNei clicked   
738292   641105  2018-04-30 07:58:59       13.0              WebNei clicked   
738299   641105  2018-04-30 07:59:12        NaN           Webnei downloaded   
738300   641105  2018-04-30 07:59:12      790.0         Category downloaded   
738399   641105  2018-04-30 08:12:22      961.0              WebNei clicked   
754009   643357  2018-05-02 09:40:59       45.0              WebNei clicked   
754010   643357  2018-05-02 09:41:44        NaN           Webnei downloaded   
754011   643357  2018-05-02 09:41:44      335.0         Category downloaded   
755866   643541  2018-05-02 11:17:41        6.0              WebNei clicked   
755870   643541  2018-05-02 11:17:47        NaN           Webnei downloaded   
755871   643541  2018-05-02 11:17:47      753.0         Category downloaded   
774479   646326  2018-05-04 08:39:51        6.0              WebNei clicked   
774481   646326  2018-05-04 08:39:57        NaN      Read document backward   
774482   646326  2018-05-04 08:39:57        3.0      Read document backward   
774484   646326  2018-05-04 08:40:00        NaN           Webnei downloaded   
774485   646326  2018-05-04 08:40:00      970.0         Category downloaded   
784421   648103  2018-05-07 08:01:00       14.0              WebNei clicked   
784422   648103  2018-05-07 08:01:14        NaN         Category downloaded   
784423   648103  2018-05-07 08:01:14      968.0           

## Possible improvements
* ### improve recommendation based on user activity during current session
* ### verification of recommendation accuracy
* ### constant flow of information about users
